grpclb_remote_balancer.go

```diff
@@ -0,0 +1,270 @@
+			if err == nil {
+				lb.subConns[addrWithoutMD] = sc // Use the addr without MD as key for the map.
+				lb.scStates[sc] = connectivity.Idle
+				sc.Connect()
+			} else {
+				grpclog.Warningf("roundrobinBalancer: failed to create new SubConn: %v", err)
+			}
+		}
+	}
+
+	for a, sc := range lb.subConns {
+		// a was removed by resolver.
+		if _, ok := addrsSet[a]; !ok {
+			backendsUpdated = true
+
+			lb.cc.RemoveSubConn(sc)
+			delete(lb.subConns, a)
+			// Keep the state of this sc in b.scStates until sc's state becomes Shutdown.
+			// The entry will be deleted in HandleSubConnStateChange.
+		}
+	}
+
+	return backendsUpdated
+}
+
+func (lb *lbBalancer) readServerList(s *balanceLoadClientStream, done chan<- struct{}) {
+	defer close(done)
+	for {
+		reply, err := s.Recv()
+		if err != nil {
+			grpclog.Errorf("grpclb: failed to recv server list: %v", err)
+			break
+		}
+		if serverList := reply.GetServerList(); serverList != nil {
+			lb.processServerList(serverList)
+		}
+	}
+}
+
+func (lb *lbBalancer) sendLoadReport(s *balanceLoadClientStream, interval time.Duration, done <-chan struct{}) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+		case <-done:
+			return
+		}
+		stats := lb.clientStats.toClientStats()
+		t := time.Now()
+		stats.Timestamp = &lbpb.Timestamp{
+			Seconds: t.Unix(),
+			Nanos:   int32(t.Nanosecond()),
+		}
+		if err := s.Send(&lbpb.LoadBalanceRequest{
+			LoadBalanceRequestType: &lbpb.LoadBalanceRequest_ClientStats{
+				ClientStats: stats,
+			},
+		}); err != nil {
+			grpclog.Errorf("grpclb: failed to send load report: %v", err)
+			return
+		}
+	}
+}
+
+func (lb *lbBalancer) watchRemoteBalancer() {
+	for {
+		select {
+		case <-lb.doneCh:
+			return
+		default:
+		}
+
+		lbClient := &loadBalancerClient{cc: lb.ccRemoteLB}
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		stream, err := lbClient.BalanceLoad(ctx, FailFast(false))
+		if err != nil {
+			grpclog.Errorf("grpclb: failed to perform RPC to the remote balancer %v", err)
+			continue
+		}
+
+		// grpclb handshake on the stream.
+		initReq := &lbpb.LoadBalanceRequest{
+			LoadBalanceRequestType: &lbpb.LoadBalanceRequest_InitialRequest{
+				InitialRequest: &lbpb.InitialLoadBalanceRequest{
+					Name: lb.target,
+				},
+			},
+		}
+		if err := stream.Send(initReq); err != nil {
+			grpclog.Errorf("grpclb: failed to send init request: %v", err)
+			continue
+		}
+		reply, err := stream.Recv()
+		if err != nil {
+			grpclog.Errorf("grpclb: failed to recv init response: %v", err)
+			continue
+		}
+		initResp := reply.GetInitialResponse()
+		if initResp == nil {
+			grpclog.Errorf("grpclb: reply from remote balancer did not include initial response.")
+			continue
+		}
+		if initResp.LoadBalancerDelegate != "" {
+			grpclog.Fatalf("TODO: Delegation is not supported yet.")
+		}
+
+		// streamDone will be closed by the readServerList goroutine when
```

Comment:
I'd like to propose an alternative to avoid passing around this channel:

- Make this a context instead.
- Cancel it here after readServerList returns.
- Pass it to sendLoadReport.

Reply:
Hmm, right, context.
There's no need to create a new context here.
`SendLoadReport()` should block on the stream's context.